src/core/net/ip6.cpp

```diff
@@ -1091,6 +1099,19 @@ Error Ip6::SendRaw(Message &aMessage, bool aAllowLoopBackToHost)
     SuccessOrExit(error = header.ParseFrom(aMessage));
     VerifyOrExit(!header.GetSource().IsMulticast(), error = kErrorInvalidSourceAddress);
 
+#if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+    // The filtering rules don't apply to packets from DUA.
+    if (!Get<BackboneRouter::Leader>().IsDomainUnicast(header.GetSource()))
+#endif
+    {
+        // When the packet is forwarded from host to Thread, if its source is on-mesh or its destination is
+        // mesh-local, we'll drop the packet unless the packet originates from this device.
+        VerifyOrExit(!(Get<NetworkData::Leader>().IsOnMesh(header.GetSource()) ||
+                       Get<Mle::Mle>().IsMeshLocalAddress(header.GetDestination())) ||
+                         Get<ThreadNetif>().GetUnicastAddresses().ContainsMatching(header.GetSource()),
+                     error = kErrorDrop);
```

Comment:
This style of `VerifyOrExit()` check is correct but rather hard to read. We have `(!(a || b) || c)` as condition. 

Adding the comment is helpful, but if we can write the code to read similar to the comment, it will be easier to follow (and we can consider removing the comment). So how about this:
```c++
        if (Get<NetworkData::Leader>().IsOnMesh(header.GetSource()) ||
            Get<Mle::Mle>().IsMeshLocalAddress(header.GetDestination()))
        {
            VerifyOrExit(Get<ThreadNetif>().HasUnicastAddress(header.GetSource()), error = kErrorDrop);
        }
```

BTW, we can use `HasUnicastAddress()` instead of `GetUnicastAddresses().ContainsMatching()`.

Reply:
Good point! I also thought it was hard to read but failed to improve the readability.